alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxow.hs

```diff
@@ -266,6 +299,88 @@ alonzoStyleWitness = do
 
   trans @(Core.EraRule "UTXO" era) $ TRC (ue, u', tx)
 
+-- | Collect the set of hashes of keys that needs to sign a given transaction.
+--  This set consists of the txin owners, certificate authors, and withdrawal
+--  reward accounts.
+--
+--  Compared to pre-Alonzo eras, we additionally gather the certificates
+--  required to authenticate collateral witnesses.
+witsVKeyNeeded ::
+  forall era tx.
+  ( Era era,
+    HasField "body" tx (Core.TxBody era),
+    HasField "wdrls" (Core.TxBody era) (Wdrl (Crypto era)),
+    HasField "certs" (Core.TxBody era) (StrictSeq (DCert (Crypto era))),
+    HasField "inputs" (Core.TxBody era) (Set (TxIn (Crypto era))),
+    HasField "collateral" (Core.TxBody era) (Set (TxIn (Crypto era))),
+    HasField "update" (Core.TxBody era) (StrictMaybe (Update era)),
+    HasField "address" (Core.TxOut era) (Addr (Crypto era))
+  ) =>
+  UTxO era ->
+  tx ->
+  GenDelegs (Crypto era) ->
+  WitHashes (Crypto era)
+witsVKeyNeeded utxo' tx genDelegs =
+  WitHashes $
```

Comment:
My policy here has always been to follow the spec, in order to make it easier to check we're doing the same thing. In this case, the spec overrides the whole `witsVKeyNeeded` function, so I did the same.

That having been said, I'm totally fine with doing this in a different way.

Reply:
That is a good policy. It is indeed that way in the spec, but only because I thought it would be weird to either rename the function to something else or to say something like `ShelleyMA.witsVKeyNeeded` in the spec. Do you think that is a better idea to rename it to a new function, and do it that way in both code and spec?